### PR TITLE
[native] Add non-blocking shuffle reader interface

### PR DESCRIPTION
**Pull request by patrickstuedi**
Avoid blocking during shuffle read to support suspending of the shuffle operator.

### CODE REVIEW

File: presto-native-execution/presto_cpp/main/operators/LocalPersistentShuffle.cpp [hunk 1]
```diff
@@ -12,6 +12,7 @@
  * limitations under the License.
  */
 #include "presto_cpp/main/operators/LocalPersistentShuffle.h"
+#include <folly/futures/Future.h>
```
Comment (yingsu00) [inline presto-native-execution/presto_cpp/main/operators/LocalPersistentShuffle.cpp:15]: nit: Add an empty line before and after

File: presto-native-execution/presto_cpp/main/operators/UnsafeRowExchangeSource.h [hunk 2]
```diff
@@ -55,5 +55,6 @@ class UnsafeRowExchangeSource : public velox::exec::ExchangeSource {
 
   // The number of batches read from 'shuffle_'.
   uint64_t numBatches_{0};
+  std::mutex mutex_;
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/operators/UnsafeRowExchangeSource.h:58]: Move mutex_ ahead of its protected fields? Thanks!

File: presto-native-execution/presto_cpp/main/operators/UnsafeRowExchangeSource.cpp [hunk 3]
```diff
@@ -32,44 +33,63 @@ folly::SemiFuture<UnsafeRowExchangeSource::Response>
...
-      atEnd_ = true;
-      queue_->enqueueLocked(nullptr, promises);
-    } else {
-      velox::BufferPtr buffer;
-      CALL_SHUFFLE(buffer = shuffle_->next(), "next");
-      totalBytes = buffer->size();
+    auto hasNext = [this]() -> folly::SemiFuture<Response> {
+      return shuffle_->hasNext()
+          .deferValue(
+              [this](bool hasNextValue) mutable
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/operators/UnsafeRowExchangeSource.cpp:40]:
```
s/hasNextValue/hasNext/
```

File: presto-native-execution/presto_cpp/main/operators/UnsafeRowExchangeSource.cpp [hunk 4]
```diff
@@ -32,44 +33,63 @@ folly::SemiFuture<UnsafeRowExchangeSource::Response>
 UnsafeRowExchangeSource::request(
     uint32_t /*maxBytes*/,
     uint32_t /*maxWaitSeconds*/) {
-  std::vector<velox::ContinuePromise> promises;
-  int64_t totalBytes = 0;
   {
-    std::lock_guard<std::mutex> l(queue_->mutex());
-    if (atEnd_) {
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/operators/UnsafeRowExchangeSource.cpp:39]: Shall we keep this?

File: presto-native-execution/presto_cpp/main/operators/UnsafeRowExchangeSource.cpp [hunk 5]
```diff
@@ -32,44 +33,63 @@ folly::SemiFuture<UnsafeRowExchangeSource::Response>
...
+                      // should have been allocated from memory pool. Hence, we
+                      // don't need to update the memory usage counting for the
+                      // associated 'ioBuf' attached to SerializedPage on
+                      // destruction.
+                      queue_->enqueueLocked(
+                          std::make_unique<velox::exec::SerializedPage>(
+                              std::move(ioBuf), [buffer](auto& /*unused*/) {}),
+                          promises);
+                    }
+                    for (auto& promise : promises) {
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/operators/UnsafeRowExchangeSource.cpp:66]: Please make sure promises set is not under the lock. Thanks!

**Review by xiaoxmeng — APPROVED** [review]
@patrickstuedi LGTM % nits. Thanks!

File: presto-native-execution/presto_cpp/main/operators/LocalPersistentShuffle.cpp (hunk 1 above)
Comment (patrickstuedi) [reply presto-native-execution/presto_cpp/main/operators/LocalPersistentShuffle.cpp:16]: Fixed. thanks!

File: presto-native-execution/presto_cpp/main/operators/UnsafeRowExchangeSource.cpp (hunk 4 above)
Comment (patrickstuedi) [reply presto-native-execution/presto_cpp/main/operators/UnsafeRowExchangeSource.cpp:39]: Added it back.

File: presto-native-execution/presto_cpp/main/operators/UnsafeRowExchangeSource.cpp (hunk 5 above)
Comment (patrickstuedi) [reply presto-native-execution/presto_cpp/main/operators/UnsafeRowExchangeSource.cpp:70]: Lock is scoped to not include setting the promise, thanks!

**Review by xiaoxmeng — APPROVED** [review]
@patrickstuedi thanks for the update!

File: presto-native-execution/presto_cpp/main/operators/LocalPersistentShuffle.cpp [hunk 6]
```diff
@@ -160,21 +160,22 @@ LocalPersistentShuffleReader::LocalPersistentShuffleReader(
...
-BufferPtr LocalPersistentShuffleReader::next() {
+folly::SemiFuture<BufferPtr> LocalPersistentShuffleReader::next() {
   if (readPartitionFiles_.empty()) {
     readPartitionFiles_ = getReadPartitionFiles();
   }
 
   if (readPartitionFileIndex_ >= readPartitionFiles_.size()) {
-    return nullptr;
+    BufferPtr buffer = nullptr;
+    return folly::makeSemiFuture<BufferPtr>(std::move(buffer));
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/operators/LocalPersistentShuffle.cpp:170]:
```
return folly::makeSemiFuture<BufferPtr>(nullptr);
```

**Review by xiaoxmeng — APPROVED** [review]
@patrickstuedi LGTM % nit. Thanks!

File: presto-native-execution/presto_cpp/main/operators/LocalPersistentShuffle.cpp (hunk 6 above)
Comment (patrickstuedi) [reply presto-native-execution/presto_cpp/main/operators/LocalPersistentShuffle.cpp]: Fixed.

**Review by xiaoxmeng — APPROVED** [review]
@patrickstuedi LGTM. Thanks!

**Pull request merged**